scrapy/spiders/__init__.py

```diff
@@ -35,11 +54,15 @@ def __init__(self, name: Optional[str] = None, **kwargs: Any):
         if name is not None:
             self.name = name
         elif not getattr(self, "name", None):
-            raise ValueError(f"{type(self).__name__} must have a name")
+            self.name = global_object_name(self.__class__)
```

Comment:
This would make spider *objects* get a name, which would solve https://github.com/scrapy/scrapy/issues/1157#issuecomment-1816782204, but spider *classes* would remain nameless.

Alternatively, we could go with the approach in https://github.com/scrapy/scrapy/pull/1649, as @GeorgeA92 [suggested](https://github.com/scrapy/scrapy/issues/1157#issuecomment-1896234239).